Add blog post: Idle Power Matters: Kepler Metrics for Public Cloud Energy Efficiency (#435)
cncf/tag-env-sustainability: website/content/en/blog/2024-06-idle-power-matters-kepler-metrics-for-public-cloud-energy-efficiency.md

However, it's important to note that cloud waste, as illustrated in situation II of the example, may still not be fully accounted for at the user level without knowledge of the requested resources of all sharing tenants.

## Conclusion

> This blog post underscores the importance of transparent, fine-grained energy metrics for public cloud efficiency. By addressing challenges such as idle power attribution and estimation, tools like Kepler pave the way for sustainable and energy-efficient cloud computing practices. However, it's crucial to acknowledge that complete waste accounting at the user level remains a challenge, necessitating further research and collaboration within the industry.

Reviewer: @sunya-ch, great blog. QQ: If this blog is about creating awareness, is there an explicit CTA? Are there links to the community channel/communication channels where interested folks can collaborate or learn more?

Author: That's very good point. I will add CTA to the blog.

- R – Functional unit

This standard encourages the use of granular data, which could be expanded to several taxonomies, including: datacenter, individual machine (e.g., VM/Node), individual service (e.g., API call or ML training run), pack of execution (e.g., pod or container), and single execution (e.g., process). 

Energy consumption represents the cumulative power usage over a period of time. Regarding power calculations for multi-tenant environments of VMs,  the [Greenhouse Gas Protocol (GHG)](https://ghgprotocol.org) provides the following specific guidelines:

<b> P(vm) = U(vm) ⋅ Size(vm) ⋅ P(dyn) + P(base)/n </b>

> This equation refers to the one found on page 6-34 of [ICT Sector Guidance built on the GHG Protocol Product Life Cycle Accounting and Reporting Standard](https://ghgprotocol.org/sites/default/files/2023-03/GHGP-ICTSG%20-%20ALL%20Chapters.pdf)`

Where:
- P(vm) - power consumption of the VM
- U(vm) - utilization of the VM
- Size(vm) - size of the VM
- P(dyn) - dynamic power range of the device
- P(base) - Idle, or base, power consumption is when the VMM is operating but no VMs are executed on the system
- n - the number of VMs running on the server

> To achieve green software building, software practitioners must collectively reduce the above power consumption and the SCI scores during the design, development, and delivery of software applications. This requires a collaborative effort to prioritize sustainability and minimize carbon emissions.

Reviewer 2: ```suggestion
To achieve greener software, we need to reduce power consumption, and the SCI scores, during the design, development, and delivery of applications. 
```

In 2019, [RightScale's survey](https://resources.flexera.com/web/media/documents/rightscale-2019-state-of-the-cloud-report-from-flexera.pdf) revealed that over 94% of 800 organizations had adopted cloud computing, with 31% prioritizing public cloud and 54% planning to migrate their workloads to the public cloud within 12 months to become more resource efficient. [Accenture's report in 2020](https://newsroom.accenture.com/news/2020/cloud-migrations-can-reduce-co2-emissions-by-nearly-60-million-tons-a-year-according-to-new-research-from-accenture) highlighted that public cloud migration can lead to a substantial reduction in carbon emissions (up to 84%) and energy consumption (up to 64%).

The public cloud operates as a multi-tenant environment where users purchase computational resources shared among numerous clients. Various tools, such as [CodeCarbon](https://codecarbon.io), [CarbonTracker](https://carbontracker.org), [Scaphandre](https://github.com/hubblo-org/scaphandre), and [Kepler](https://sustainable-computing.io), are available for estimating energy consumption and carbon emissions associated with code execution wrapping or real-time resource utilization observation.

> However, determining how to attribute power consumption for infrastructure services presents a challenge. The GSF's SCI standard specifies that energy consumption should encompass all energy consumed by reserved or provisioned hardware, not solely the hardware utilized for application software needs. This includes idle power, encompassing not only static power but also the energy consumed by control plane processes such as kubelet or container runtime (e.g., containerd) in kubernetes orchestration contexts.

Reviewer 2: ```suggestion
However, determining how to attribute power consumption for infrastructure services remains challenging. The GSF's SCI standard specifies that energy consumption should encompass all energy consumed by reserved or provisioned hardware, not solely the hardware used while an application is running. This includes idle power, encompassing not only static power but also the energy consumed by items necessary to be functional for the application to run such as the kubelet or the control plane.
```

> According to the power data reported by [SPECpower benchmark](https://www.spec.org/power_ssj2008/), active idle power consumption can vary widely, ranging from 20% to 60% of the power consumed at maximum utilization. In absolute terms, this ranges from 100Watt to 1300Watt, equivalent to the power consumption of a classic light bulb or an air conditioner in a household (although there are differences in carbon emissions). The idle power can be even higher when considering GPU power consumption. Therefore, idle power is significant and must be allocated to each application. Then, the challenge lies in determining how to do so effectively.

Reviewer 2: ```suggestion
According to the [SPECpower benchmark](https://www.spec.org/power_ssj2008/), active idle power consumption can vary widely, ranging from 20% to 60% of the power consumed at maximum utilization. In absolute terms, this ranges from 100Watts to 1300Watts, equivalent to the power consumption of a classic light bulb or an air conditioner in a household (although there are differences in carbon emissions). The idle power can be even higher when considering GPU power consumption.
```

Removed a couple of sentences-not sure if they add to the paragraph significantly.

## The Importance of Idle power accounting and Transparency

> The differentiation between idle and dynamic power has been widely investigated in the research literature [[1]](https://ieeexplore.ieee.org/document/9139675). However, most current tools for measuring application power consumption often neglect this distinction. It is crucial to isolate idle power from dynamic power for several reasons. Not only is it important for fair attribution of power consumption to applications, but it is also essential for creating more accurate power models.

Reviewer 2: ```suggestion
The differentiation between idle and dynamic power has been widely investigated in the research literature [[1]](https://ieeexplore.ieee.org/document/9139675). Most current tools for measuring application power consumption often neglect this distinction. It is crucial to isolate idle power from dynamic power for reasons including fair attribution of power consumption to applications and to help create more accurate power models.
```

> Dynamic energy consumption is directly related to resource usage. On the other hand, idle power is the base power consumption that occurs regardless of resource utilization. Therefore, a power model that uses resource utilization to estimate power consumption will only accurately reflect dynamic power. Idle power, on the other hand, is a constant power consumption that must be distributed among all running applications.

Reviewer 2: ```suggestion
Dynamic energy consumption is directly correlated to resource use. On the other hand, idle power is the base power consumption that occurs regardless of resource utilization. A power model that uses resource utilization to estimate power consumption will only accurately reflect dynamic power. Idle power, is a constant power consumption that must be attributed to all running applications.
```

> To better illustrate this concept of fair distribution of idle energy, consider the scenario of a condominium where owners must pay a homeowners association (HOA) fee whether they use their homes or not. This rate is analogous to idle energy. Additionally, larger homes may incur higher fees, similar to how virtual machines (VMs) with more allocated resources or applications with more processes should have a greater share of idle power consumption.

Reviewer 2: ```suggestion
To illustrate the idea of fairly distributing idle energy, think of a condominium where owners pay a homeowners association (HOA) fee regardless of whether they use their homes. Larger homes in the condo might pay higher fees, just as virtual machines (VMs) with more resources allocated or applications with more processes should bear a greater portion of idle power usage.
```

> By properly accounting for idle and dynamic energy, [Kepler](https://sustainable-computing.io) can ensure a fairer distribution of power consumption and develop more accurate power consumption models. In Kepler the definitions go even further, the aggregated power consumption of idle and dynamic power is called absolute power consumption.

Reviewer 2: ```suggestion
Kepler, through careful tracking of idle and dynamic energy, promotes a fairer distribution of power usage and refines power consumption models. In Kepler, the aggregate  of the consumption of idle and dynamic power is referred to as absolute power consumption.
```

> More specifically, to further illustrate the importance of transparency in idle power attribution, consider the following example. `By resource request` refers to the [OpenCost specification](https://www.opencost.io/docs/specification) where workload costs are computed by `max(request, usage)`. 

Reviewer 2: ```suggestion
To further illustrate the importance of transparency in idle power attribution, refer to the following example. `By resource request` refers to the [OpenCost specification](https://www.opencost.io/docs/specification) where workload costs are computed by `max(request, usage)`. 
```

<p class="mt-5 mb-5"><img src="/images/blogs/2024-06-idle-power-matters-kepler-metrics-for-public-cloud-energy-efficiency/idle_power_blog_example.webp" alt="Example of Power Attribution in Multi-tenant Environments"></p>

Suppose we have a server with a linear power model for a 16-core server X, described by a simple formula as simple as below where Idle=200 and Coeff=8:

<b>Power Consumption (P) = Idle Power (Idle) + Dynamic Power Consuming Factor (Coeff) * Resource Utilization (x)</b>

> Let's analyze the estimated energy consumption for the situations illustrated in the figure above. Consider a server with two clients/applications requesting the same amount of resources but with different resource utilizations in different situations. One situation will illustrate a single client for comparison.

Reviewer 2: ```suggestion
Let's analyze the estimated energy consumption shown in the figure above. We will consider a server hosting two clients or applications that request the same amount of resources but utilize them differently in various scenarios. We'll also examine a scenario with just one client for comparison.
```

- Situation III - new client with higher resource utilization

> Now consider a third situation with uneven resource utilization: Client A still uses 10%, but Client B now uses 40% of the node's CPU. If we only use the node's absolute power consumption of 600W, Client B's resource utilization rate of 80% results in a power consumption of 480W and 120W for Client A. However, If we isolate the idle power of 200W, Client B's power consumption will be 420W (320W for dynamic + 100W for idle) and Client A's power consumption will be 180W (80W for dynamic as it is still using 10% of the CPU + 100W for idle).

- Situation IV - new client with lower resource utilization

> The situation IV is similar to the third situation, but Client B is using fewer resources than Client A, further illustrating the potential for unfair distribution of idle power.

> As we can see In the third scenario, isolating the idle power makes the results more predictable and easier to understand. This approach fairly divides the constant power (which does not vary with resource consumption) and defines the dynamic power entirely based on resource utilization. In more details, using the power model, we know that Client A's dynamic power is 80W. If we allocate idle power only using the resource ration and absolute power, Client A receives only 40W of idle power, leaving Client B with 160W of idle power.

Reviewer 2: ```suggestion
In the third scenario, isolating idle power enhances predictability and clarity. This method divides constant power (unaffected by resource use) and defines dynamic power solely based on resource utilization. Specifically, according to the power model, Client A's dynamic power is 80W. By allocating idle power based on resource ratio and absolute power, Client A receives 40W of idle power, while Client B gets 160W. and absolute power, Client A receives only 40W of idle power, leaving Client B with 160W of idle power.
```

> These unequal distributions of idle power highlights the unfairness, as idle power consumption is not related to resource utilization. On the other hand, applying the power consumption ratio to resource request does not reflect the real dynamic power consumption according to the real resource utilization. 

Reviewer 2: ```suggestion
These unequal distributions of idle power highlight unfairness because idle power consumption isn't tied to resource utilization. Applying the power consumption ratio based on resource requests doesn't accurately reflect actual dynamic power consumption corresponding to real resource utilization.
```

> Furthermore, it is challenging to justify why more idle power should be associated with one client over another purely based on resource utilization. As mentioned before, a fair distribution of idle power would consider the application size or simply distribute the idle power evenly among all applications.

Reviewer 2: ```suggestion
It is difficult to justify giving more idle power to one client over another solely based on resource utilization. A fair distribution of idle power could either account for application size or distribute idle power cost evenly among all applications.
```

## The Challenges of Estimating Idle Power

Now that we understand the importance of isolating and reporting idle power, let's delve into how Kepler addresses the challenge of estimating idle power, especially in cloud environments where certain information, such as the number of co-located virtual machines, may be obscured.

On bare metal, we can estimate idle power by identifying the minimum point of power consumption when server utilization is close to zero [2, 3]. However, we plan to extend this model with regression analysis to find the idle power also in busy systems.

> However, estimating idle power becomes more complex in virtual private clouds, where physical specifications and the number of sharing tenants (virtual machines) are not readily available. To address this, we've developed a training framework to extract and train only the dynamic power model, thereby reducing dependency on unknown factors and providing energy-efficient optimization for virtual machines [[4]](https://www.computer.org/csdl/proceedings-article/mascots/2023/10387542/1TKR5vzaWXe).

Reviewer 2: ```suggestion
Estimating idle power in virtual private clouds is challenging due to unknown physical specs and the number of tenants (VMs). To address this, we've developed a training framework focused on extracting and training the dynamic power model alone. This approach reduces dependency on unknown factors and enhances energy-efficient optimization for virtual machines [[4]](https://www.computer.org/csdl/proceedings-article/mascots/2023/10387542/1TKR5vzaWXe).
```

> At the beginning, as addressed in our previous blog post [[5]](https://www.cncf.io/blog/2023/10/11/exploring-keplers-potentials-unveiling-cloud-application-power-consumption/), we decided to disable the idle power of the container running on top of VM. Yet, we still need to tackle the idle power attribution and accounting issue. To strike a balance, we're developing an approach to estimate the idle power of virtual machines based on limited knowledge, specifically in scenarios where (i) physical cores are pinned to virtual cores without overprovisioning, and (ii) the same CPU architecture is emulated in the virtual machine. In such cases, we can apply a power model trained on the same or similar CPU architecture and estimate the idle power specific to the virtual machine based on the number of cores. For instance, if we have a power model trained for 64 cores with an idle power of 160W, a virtual machine provided with 8 virtual cores would account for 20W (i.e. 160/8).

Reviewer 2: ```suggestion
As discussed in our previous blog post [[5]](https://www.cncf.io/blog/2023/10/11/exploring-keplers-potentials-unveiling-cloud-application-power-consumption/), we disabled idle power for containers running on VMs. We still need to address the attribution and accounting of idle power. To find a balance, we're developing an approach to estimate VM idle power based on limited information, particularly in scenarios where physical cores are pinned to virtual cores without overprovisioning and the VM emulates the same CPU architecture. For these, we can use a power model trained on similar CPU architectures to estimate VM idle power based on core count. For example, if our model for 64 cores predicts an idle power of 160W, a VM with 8 virtual cores would estimate around 20W (i.e., 160/8).
```

However, it's important to note that cloud waste, as illustrated in situation II of the example, may still not be fully accounted for at the user level without knowledge of the requested resources of all sharing tenants.

## Conclusion

> This blog post underscores the importance of transparent, fine-grained energy metrics for public cloud efficiency. By addressing challenges such as idle power attribution and estimation, tools like Kepler pave the way for sustainable and energy-efficient cloud computing practices. However, it's crucial to acknowledge that complete waste accounting at the user level remains a challenge, necessitating further research and collaboration within the industry.

Reviewer 2: ```suggestion
Complete waste accounting at the user level remains a challenge, necessitating further research and collaboration within the industry.  Transparent, fine-grained energy metrics for public cloud efficiency remains important, and needs to have challenges addressed such as idle power attribution and estimation. Tools like Kepler pave the way for sustainable cloud computing practices. 
```
